newIDE/app/src/MainFrame/index.js

```diff
@@ -853,54 +840,72 @@ class MainFrame extends React.Component<Props, State> {
-          eventsFunctionsExtension
-        ),
-      },
-      () => {
-        // Refactor the project to update the instructions (and later expressions)
-        // of this extension:
-        gd.WholeProjectRefactorer.renameEventsFunctionsExtension(
-          currentProject,
-          eventsFunctionsExtension,
-          oldName,
-          newName
-        );
-        eventsFunctionsExtension.setName(newName);
-        eventsFunctionsExtensionsState.reloadProjectEventsFunctionsExtensions(
-          currentProject
-        );
-        this._onProjectItemModified();
-      }
-    );
+    setState(state => ({
+      ...state,
+      editorTabs: closeEventsFunctionsExtensionTabs(
+        state.editorTabs,
+        eventsFunctionsExtension
+      ),
+    })).then(state => {
+      // Refactor the project to update the instructions (and later expressions)
+      // of this extension:
+      gd.WholeProjectRefactorer.renameEventsFunctionsExtension(
+        currentProject,
+        eventsFunctionsExtension,
+        oldName,
+        newName
+      );
+      eventsFunctionsExtension.setName(newName);
+      eventsFunctionsExtensionsState.reloadProjectEventsFunctionsExtensions(
+        currentProject
+      );
+      _onProjectItemModified();
+    });
   };
 
-  _launchLayoutPreview = (
+  const [_layoutPreview, _setLayoutPreview] = React.useState({
+    project: null,
+    layout: null,
+    options: null,
+  });
+
+  const _launchLayoutPreview = (
     project: gdProject,
     layout: gdLayout,
     options: PreviewOptions
   ) => {
-    const { _previewLauncher } = this;
-    const { previewFirstSceneName, isPreviewFirstSceneOverriden } = this.state;
+    const { previewFirstSceneName, isPreviewFirstSceneOverriden } = state;
 
-    if (!_previewLauncher) return;
+    if (!_previewLauncher.current) return;
 
-    this.setState(
-      {
-        previewLoading: true,
-      },
-      () => {
-        let previewedLayout = layout;
-        if (previewFirstSceneName && isPreviewFirstSceneOverriden) {
-          if (project.hasLayoutNamed(previewFirstSceneName)) {
-            previewedLayout = project.getLayout(previewFirstSceneName);
-          }
-        }
+    setState(state => ({
+      ...state,
+      previewLoading: true,
+    }));
 
-        _previewLauncher
+    _setLayoutPreview({
+      project: project,
+      layout: layout,
+      options: options,
+    });
+    //callback shifted to useEffect
+  };
+
+  React.useEffect(
+    () => {
+      const previewedLayout = _layoutPreview.layout;
+      const project = _layoutPreview.project;
+      const options = _layoutPreview.options;
+      if (state.previewFirstSceneName && state.isPreviewFirstSceneOverriden) {
+        if (project && project.hasLayoutNamed(state.previewFirstSceneName)) {
+          _setLayoutPreview(layoutPreview => ({
+            ...layoutPreview,
+            layout: project.getLayout(state.previewFirstSceneName),
+          }));
+        }
```

Comment:
To fix the infinite loop, when choosing a scene to be used as the first scene for all previews (see my other comment explaining how to reproduce the bug), I'm pretty sure this should be moved to `_launchLayoutPreview `.
There should be no calls to `_setLayoutPreview` in the effect that is responsible for... reacting to changes made in `_setLayoutPreview` ;)

Reply:
I've fixed this in the latest commits, give it a try once again and tell me if it's working properly :)